### PR TITLE
networkd: support marking links unmanaged

### PR DESCRIPTION
**Pull request by dm0-**
This is a request to allow explicitly specifying that links are not to be managed by networkd.  To get this effect currently, all network files have to be structured to avoid matching the desired interfaces.  (Having an earlier-matching network file with no settings is not enough; that still manages the interface which wipes configured addresses, link master, etc.)  This change adds an `Unmanaged` boolean to the `Link` section to address this.

The primary motivation for this is that networkd breaks Docker's default networking driver (at least when there are matching network files).  It results in race conditions between Docker configuring its newly created veth/bridge interfaces and networkd wiping the existing configuration on new interfaces.  As a distributor, we would like to be able to ship high priority network files that match such interfaces and tell networkd not to interact with them.  This would help to prevent accidentally matching these interfaces by end-users' network files and any overly inclusive distro network files.

### CODE REVIEW

File: man/systemd.network.xml [hunk 1]
```diff
@@ -232,6 +232,20 @@
...
       </varlistentry>
+      <varlistentry>
+        <term><varname>Unmanaged=</varname></term>
+        <listitem>
+          <para>A boolean. When <literal>yes</literal>, no attempts are
+          made to bring up or configure matching links, equivalent to
+          when there are no matching network files. Defaults to
+          <literal>no</literal>.</para>
+          <para>This is useful for preventing later matching network
+          files from interfering with certian interfaces that are fully
```
Comment (martinpitt) [inline man/systemd.network.xml]: "certain"

File: man/systemd.network.xml [hunk 2]
```diff
@@ -232,6 +232,20 @@
...
+          controlled by other applications, such as virtual Ethernet
+          devices that are created and managed internally by sandboxing
+          applications.</para>
```
Comment (martinpitt) [inline man/systemd.network.xml]: I'm afraid that "such as" explanation does not really explain anything. Can you try to rephrase that, or leave it out?

**Comment by johannbg** [comment]
@martinpitt  this looks like dockers default networking driver is broken. @dm0-  is there any bug report against docker we can have a look at to see what the actual issue people are experiencing?

**Comment by dm0-** [comment]
@martinpitt I've updated the doc with your suggestions.  I'll look into the test later today.

I think it is useful to have a blacklist/exclude type of setting in this case because there is no reason for networkd to ever interfere with these interfaces.  They are completely managed by Docker.

A different solution that might help matching would be to support negation in the `Driver=` field, so something like `Driver=!veth` can be applied to broad matches in late `.network` files.  (I wouldn't prefer this since it seems to say "manage everything except Docker interfaces", as opposed to `Unmanaged=yes` saying the stronger statement "don't manage Docker interfaces", which is more robust in preventing later accidental matching.)

Another alternative would maybe be `DontWipeExistingConfiguration=yes`, but I don't know how different that would be from `Unmanaged=yes` in practice.

@johannbg Yes, the main Docker issue that I responded to is docker/docker#26492.  When I first responded, I wasn't aware networkd was involved and was only looking at netlink events in Docker.

I don't think Docker is doing anything unreasonable here.  They are just creating veth devices, configuring them, and adding them to the bridge, not expecting any external service to interfere with them.  Then, networkd wipes their configuration and removes them from the bridge when any `.network` file matches, leaving the containers with no networking.

**Comment by marineam** [comment]
@martinpitt and a little more context, the catch-all is used in CoreOS to enable DHCP by default. Also, the kernel doesn't provide a "ethernet" or similar DEVTYPE so there wasn't really a practical way to write a more restrictive default.

**Comment by dm0-** [comment]
I've written a test class, but it's completely untested--as in I haven't even attempted to execute it yet.  It's uploaded now so you can comment on the idea, but I'll continue working on it later when I have a working testing environment.

**Comment by dm0-** [comment]
I've pushed a minor encoding fix to the test script, but otherwise, the tests I added worked in a Fedora 25 container.  (I'd question whether the base client class should define tests to be inherited by the other classes, but I left that as-is and declared the new class to be semi-consistent with the others.)  Any other suggestions?

**Comment by martinpitt** [comment]
I'm afraid this is untestable -- the new test somehow manages to break the connection of the main real `ens2` interface so that the test controller loses ssh to the testbed. I salvaged the `networkd-test.py` output until the disconnection error: http://paste.ubuntu.com/23278454/ (not sure how helpful that actually is).

In this context, `testbed auxverb failed with exit code 255` means the exit code of `ssh`, where 255 is an unexpected disconnect.

**Comment by dm0-** [comment]
@martinpitt Sorry about that.  From where the output ends, it looks like the test's networkd matched your interface and starting managing it, wiping its existing configuration.  I.e. the test triggered the issue the patch aims to help prevent.  (I missed that myself because I tested it inside `systemd-nspawn --private-network ...`.)  I've amended the commit to use a stricter match for that test to hopefully not mess with the test system's interfaces, and to also skip redundant inherited tests in the new class.

**Comment by poettering** [comment]
Hmm, instead of adding this option, I wonder if it wouldn't be better to really add negative matches, i.e. Driver=~veth, so that it is easy to write catch-all .network files that exclude veth links. This would match the negative conditions we already expose in service files. 

Would that work for you? Any chance you can send a patch for that?

**Comment by dm0-** [comment]
@poettering Supporting negative conditions would certainly help, but I don't think it's ideal.
- Negated conditions alone aren't enough to express inverted `[Match]` sections with multiple lines; trigger conditions would also be needed to have an "or" relation.
- End users writing their own network files may need to be aware of all the excluded matches in the catch-all files so they don't hit accidentally match them.
- If any users need to exclude an interface not considered by the distro's catch-all files, they would have to override all of them as opposed to just writing a new `Unmanaged=` file.

That being said, if a negative condition is the only acceptable option, I've written a new function that checks if the pattern strv begins with `!`.  It copies the strv to edit out the `!` before the `fnmatch` without modifying the memory passed to the function, and returns the negated match result.

I've cherry-picked the `Unmanaged` commit onto the end in case anyone else wants to comment on it.  It can be dropped if it is not going to be used.

File: test/networkd-test.py [hunk 3]
```diff
@@ -594,6 +594,162 @@ def test_dropin(self):
         self.assertIn('nameserver 192.168.42.1\n', contents)
         self.assertIn('nameserver 127.0.0.1\n', contents)
 
+
+class UnmanagedClientTest(ClientTestBase, unittest.TestCase):
```
Comment (martinpitt) [inline test/networkd-test.py]: Please don't subclass `ClientTestBase` for adding these tests -- this is meant to test networkd (client side) against different "remote" servers (dnsmasq or networkd in DHCP server mode), and all these `skip("Inherited")` are ugly.

Either add the new tests to the existing `NetworkdClientTest` class (no need to run them for dnsmasq too), or keep it as a new class but directly derive from `unittest.TestCase`. If there is much functionality that you need from `ClientTestBase` it is also fine to only keep the utility functions in there, and move the actual shared tests into an intermediate `CommonTests(ClientTestBase)` class and have `{Dnsmasq,Networkd}ClientTests` inherit from `CommonTests`.

That said, no need to change the tests before agreeing to the actual design of course (that still seems to be disputed).

Comment (dm0-) [reply test/networkd-test.py]: I've added another commit that makes a separate utility class that I've added to.  (Sorry for the delay in that.)  More functionality could probably be migrated to it in a later pass.

**Comment by dm0-** [comment]
I've pushed an update to address conflicts from recent networkd test additions.

**Comment by lucab** [comment]
Ping @poettering @martinpitt @keszybz.

How do we want to proceed from here? Is there some consensus on which approach is preferred?

I tend to agree with @dm0- in preferring an `Unmanaged=` properties over negative matches. However, outside of this docker bug, it looks like those may be two orthogonal features with other different usecases. In the light of https://github.com/systemd/systemd/issues/3998, should we perhaps split negative-match out of here into its own PR for further discussion?

File: test/networkd-test.py [hunk 4]
```diff
@@ -39,18 +40,70 @@
...
+        self.addCleanup(os.remove, dropin_path)
+
+
+class ClientTestBase(NetworkdTestingUtilities):
+    """Provide common methods for testing networkd against servers."""
+
     @classmethod
     def setUpClass(klass):
+        if shutil.which('journalctl') is None:
+            raise OSError((errno.ENOENT, 'journalctl not found'))
```
Comment (keszybz) [inline test/networkd-test.py]: Too many parentheses here.

**Comment by keszybz** [comment]
I agree that this sounds like two separate features. Actually both sound useful in general... but to solve this particular case, where we want to mark a subset of interfaces as unmanaged so that they can be controlled externally, negative matches are not useful, because *all* matching would have to be modified. Explicitly marking a link as unmanaged is much easier and more robust.

About "prior art": NetworkManager has `NM_CONTROLLED=yes|no`, which is extremely useful in some context. I think we should do the same.

So, I'd propose to: drop patch 3, fix the extra parenthesis I pointed out above, and merge.

We can then revisit patch 3 in a separate PR. At least it'll need some additions to documentation.

File: test/networkd-test.py (hunk 4 above)
Comment (dm0-) [reply test/networkd-test.py]: I've pushed an amended commit that updated the four `OSError`s to match the Python 3 constructor style.

**Comment by keszybz** [comment]
LGTM. @poettering, @martinpitt, ok to merge?

File: test/networkd-test.py [hunk 5]
```diff
@@ -39,18 +40,70 @@
...
 
-networkd_active = subprocess.call(['systemctl', 'is-active', '--quiet',
-                                   'systemd-networkd']) == 0
-have_dnsmasq = shutil.which('dnsmasq')
+HAVE_DNSMASQ = shutil.which('dnsmasq') is not None
+
+NETWORK_UNITDIR = '/run/systemd/network'
+
+NETWORKD_WAIT_ONLINE = shutil.which('systemd-networkd-wait-online',
+                                    path='/usr/lib/systemd:/lib/systemd')
```
Comment (martinpitt) [inline test/networkd-test.py:48]: capital-letter variables are very unpythonic (https://www.python.org/dev/peps/pep-0008/#id42).

File: test/networkd-test.py [hunk 6]
```diff
@@ -39,18 +40,70 @@
...
 
-@unittest.skipIf(networkd_active,
-                 'networkd is already active')
-class ClientTestBase:
+def setUpModule():
+    """Initialize the environment, and perform sanity checks on it."""
+    if shutil.which('systemctl') is None:
+        raise OSError(errno.ENOENT, 'systemctl not found')
+    if shutil.which('networkctl') is None:
+        raise OSError(errno.ENOENT, 'networkctl not found')
```
Comment (martinpitt) [inline test/networkd-test.py]: This seems a bit redundant to me, TBH -- tests will fail with a CalledProcessError anyway if these aren't present.

File: test/networkd-test.py [hunk 7]
```diff
@@ -39,18 +40,70 @@
...
+NETWORKD_WAIT_ONLINE = shutil.which('systemd-networkd-wait-online',
+                                    path='/usr/lib/systemd:/lib/systemd')
 
 RESOLV_CONF = '/run/systemd/resolve/resolv.conf'
 
 
-@unittest.skipIf(networkd_active,
-                 'networkd is already active')
-class ClientTestBase:
+def setUpModule():
```
Comment (martinpitt) [inline test/networkd-test.py:53]: Nice, I learned about the existence of `setUpModule()`. Thanks!

File: test/networkd-test.py [hunk 8]
```diff
@@ -39,18 +40,70 @@
...
+        self.addCleanup(os.remove, dropin_path)
+
+
+class ClientTestBase(NetworkdTestingUtilities):
+    """Provide common methods for testing networkd against servers."""
+
     @classmethod
     def setUpClass(klass):
+        if shutil.which('journalctl') is None:
+            raise OSError(errno.ENOENT, 'journalctl not found')
```
Comment (martinpitt) [inline test/networkd-test.py]: Again, I find these checks redundant. They just make the failure slightly different, and the chance of actually hitting it is very small anyway?

File: test/networkd-test.py [hunk 9]
```diff
@@ -39,18 +40,70 @@
...
+
+
+class NetworkdTestingUtilities:
+    """Provide a set of utility functions to facilitate networkd tests.
+
+    This class must be inherited along with unittest.TestCase to define
+    some required methods.
+    """
+
+    def write_network_unit(self, unit_name, contents):
```
Comment (martinpitt) [inline test/networkd-test.py]: Nitpick: These aren't really units, so this is a bit confusing. `write_network()` seems quite alright?

File: test/networkd-test.py [hunk 10]
```diff
@@ -96,6 +96,46 @@ def write_network_dropin(self, unit_name, dropin_name, contents):
...
+        before the check is performed.  Every specified interface must
+        be present in the output, and any other interfaces found in the
+        output are ignored.
+
+        A special interface state "managed" is supported, which matches
+        any value in the "SETUP" column other than "unmanaged".
+        """
+        if not kwargs:
+            return
+        succeeded = dict.fromkeys(kwargs.keys(), False)
```
Comment (martinpitt) [inline test/networkd-test.py]: This and the tests below could be simplified a bit by replacing this with `interfaces = set(kwargs)`...

File: test/networkd-test.py [hunk 11]
```diff
@@ -96,6 +96,46 @@ def write_network_dropin(self, unit_name, dropin_name, contents):
...
+            fields = line.split()
+            if len(fields) >= 5 and fields[1] in kwargs:
+                iface = fields[1]
+                expected = kwargs[iface]
+                actual = fields[-1]
+                if (actual != expected and
+                        not (expected == 'managed' and actual != 'unmanaged')):
+                    self.fail("Link %s expects state %s, found %s" %
+                              (iface, expected, actual))
+                succeeded[iface] = True
```
Comment (martinpitt) [inline test/networkd-test.py]: `interfaces.remove(iface)`...

File: test/networkd-test.py [hunk 12]
```diff
@@ -96,6 +96,46 @@ def write_network_dropin(self, unit_name, dropin_name, contents):
...
+
+        # Ensure that all requested interfaces have been covered.
+        if not all(succeeded.values()):
```
Comment (martinpitt) [inline test/networkd-test.py]: `if interfaces:`

File: test/networkd-test.py [hunk 13]
```diff
@@ -96,6 +96,46 @@ def write_network_dropin(self, unit_name, dropin_name, contents):
...
+            self.fail("Missing links in status output: %s" %
+                      [i for i in succeeded if not succeeded[i]])
```
Comment (martinpitt) [inline test/networkd-test.py]: and just printing interfaces here.

File: test/networkd-test.py [hunk 14]
```diff
@@ -726,6 +766,83 @@ def get_tz():
...
+        # Define the contents of .network files to be read in order.
+        self.configs = (
+            "[Match]\nName=m1def\n",
+            "[Match]\nName=m1unm\n[Link]\nUnmanaged=yes\n",
+            "[Match]\nName=m1*\n[Link]\nUnmanaged=no\n",
+        )
+
+        # Write out the .network files to be cleaned up automatically.
+        for i in range(len(self.configs)):
+            self.write_network_unit("%02d-test.network" % i, self.configs[i])
```
Comment (martinpitt) [inline test/networkd-test.py]: This iteration is a bit unwieldy.. `for c in self.configs:` and just pass `c` as the value.

File: test/networkd-test.py [hunk 15]
```diff
@@ -726,6 +766,83 @@ def get_tz():
...
+        for peer in self.veths.values():
+            subprocess.check_call(['ip', 'link', 'del', 'dev', peer])
+        subprocess.call(['systemctl', 'stop', 'systemd-networkd'])
+
+    def create_iface(self):
+        """Create some veth pairs to use for interface matching."""
+        for veth in self.veths:
+            subprocess.check_call(['ip', 'link', 'add',
+                                   'name', veth, 'type', 'veth',
+                                   'peer', 'name', self.veths[veth]])
```
Comment (martinpitt) [inline test/networkd-test.py]: It's a bit more compact and easier to see what happens to put a `self.addCleanup(subprocess.call, [..stop..]` here and drop it from `tearDown()`.

File: test/networkd-test.py (hunk 5 above)
Comment (keszybz) [reply test/networkd-test.py:48]: Actually globals are supposed to be written in uppercase.

Comment (keszybz) [reply test/networkd-test.py:48]: OK, I see that the PEP 8 recommends lower case. But I'm pretty sure that there was a recommendation to use uppercase at some point.

File: test/networkd-test.py (hunk 6 above)
Comment (keszybz) [reply test/networkd-test.py]: Yeah, seems better to remove those checks.

File: test/networkd-test.py (hunk 5 above)
Comment (dm0-) [reply test/networkd-test.py:48]: Running pylint on the file considers those variables to be constants, so it complains when they are not in caps-with-underscore naming format.  https://www.python.org/dev/peps/pep-0008/#constants

I'll make the other changes you suggest but leave this as is (unless you confirm these should go against pylint's detection before I push the amended commit).

Comment (keszybz) [reply test/networkd-test.py:48]: I don't think we care about pylint too deeply, but you're right, UPPERCASE is closer to the guidelines here.

Comment (martinpitt) [reply test/networkd-test.py:48]: Yeah, let's not get drawn into a bikeshed argument here -- just keep it as it is now.

File: test/networkd-test.py (hunk 14 above)
Comment (dm0-) [reply test/networkd-test.py]: I've made all the other changes, but this one is slightly different.  I've looped over the result of `enumerate` to get the index instead, which seems to be the most Pythonic way to do this.  I think using the list index here is necessary to properly number the `.network` files for the test.

Comment (martinpitt) [reply test/networkd-test.py]: @dm0-: Oh right, nevermind me, I was overlooking the `% i`.

**Comment by keszybz** [comment]
```
curl: (7) Failed connect to copr.fedorainfracloud.org:80; No route to host
```

**Comment by dm0-** [comment]
Thanks, everyone, for helping with this.  I've created #4809 for the commit that was dropped from this PR to invert `[Match]` lines, if there is still interest in adding that as a separate feature.

**Pull request merged**